File: leetcode/medium/633. Sum of Square Numbers.ts
function judgeSquareSum(c: number): boolean {
    const sqrt_c: number = Math.floor(Math.sqrt(c));

    // TYPE 1
    // faster solution : using hash table
    const sqrt_obj: {[num_sqr: number]: boolean} = {};
    for (let num_1 = 0; num_1 <= sqrt_c; num_1 += 1){
        sqrt_obj[num_1**2] = true;
    }
    // linear on part 1 solution
    for (let num_2 = 0; num_2 <= sqrt_c; num_2 += 1){
        const sqr_2: number = num_2**2;
        const diff: number = c - sqr_2;
        // if (sqr_2 === diff) continue;
        if (sqrt_obj[diff]) return true;
    }
    return false;

    // TYPE 2
    // slow solution
    for (let num_1 = 0; num_1 <= sqrt_c; num_1 += 1){
        for (let num_2 = num_1 + 1; num_2 <= sqrt_c; num_2 += 1){
            if ((num_1**2 + num_2**2) === c) return true;
        }
    }
    return false;
};

File: leetcode/medium/633. Sum of Square Numbers.rs
use std::collections::HashMap;
impl Solution {
    pub fn judge_square_sum(c: i32) -> bool {
        let sqrt_c: i64 = (c as f64).sqrt() as i64;
        // faster solution
        let mut start: i64 = 0;
        let mut end: i64 = sqrt_c;

        while start <= end{
            let sum_square: i64 = start*start + end*end;
            if sum_square > c as i64{
                end -= 1;
                continue;
            }
            if sum_square < c as i64{
                start += 1;
                continue;
            }
            return true;
        }
        return false;

        // slow solution
        let mut sqr_1: HashMap<i32, i32> = HashMap::new();
        for num_1 in 0..(sqrt_c + 1) as i32{
            sqr_1.insert(num_1*num_1, num_1);
        }
        for num_2 in 0..(sqrt_c + 1) as i32{
            let diff_sqr_2: i32 = c - (num_2*num_2);
            if sqr_1.contains_key(&diff_sqr_2){
                return true;
            }
        }
        return false;
    }
}
